{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This is a library of functions for working \n",
    "with fastq files and iBest software.\n",
    "\"\"\"\n",
    "from __future__ import print_function\n",
    "import sys, os, subprocess\n",
    "import string\n",
    "import numpy as np\n",
    "import random\n",
    "import pandas as pd\n",
    "from Bio import pairwise2\n",
    "import fnmatch\n",
    "import filecmp\n",
    "\n",
    "\n",
    "# find all files, output names\n",
    "def find_fastq_files(directory, pattern):\n",
    "    \"\"\"Walks the directory structure, appending filenames to an array\"\"\"\n",
    "    filenames = []\n",
    "    for root, dirs, files in os.walk(directory):\n",
    "        for basename in files:\n",
    "            if fnmatch.fnmatch(basename, pattern):\n",
    "                filename = os.path.join(root, basename)\n",
    "                filenames.append(filename)\n",
    "    filenames.sort()\n",
    "    return filenames\n",
    "\n",
    "\n",
    "# basic call to the shell application we are testing\n",
    "def sub_process(command):\n",
    "    return subprocess.check_output(\n",
    "        command, stderr=subprocess.STDOUT, shell=True)\n",
    "\n",
    "def file_compare(command, expected, returned):\n",
    "    # testing expected file output\n",
    "    subprocess.check_output(command, stderr=subprocess.STDOUT, shell=True)\n",
    "    return filecmp.cmp(expected, returned)\n",
    "\n",
    "\n",
    "def parse_fastq(filename):\n",
    "    # output the file to a dictionary\n",
    "    with open(filename) as f:\n",
    "        lines = f.readlines()\n",
    "    head = [item[:-1] for item in lines[0::4]]\n",
    "    read = [item[:-1] for item in lines[1::4]]\n",
    "    qual = [item[:-1] for item in lines[3::4]]\n",
    "    return {'Header': head, 'Sequence': read, 'QScore': qual}\n",
    "\n",
    "def word_count(wordstring):\n",
    "    wordlist = list(wordstring)\n",
    "\n",
    "    wordfreq = []\n",
    "    for w in wordlist:\n",
    "        wordfreq.append(wordlist.count(w))\n",
    "\n",
    "    print (\"String\" + \"\\n\" + wordstring +\"\\n\")\n",
    "    print (\"List\" + \"\\n\" + str(wordlist) + \"\\n\")\n",
    "    print (\"Frequencies\" + \"\\n\" + str(wordfreq) + \"\\n\")\n",
    "    print (\"Pairs\" + \"\\n\" + str(zip(wordlist, wordfreq)))\n",
    "    \n",
    "def gen_sequence(length):\n",
    "    \"\"\"Return a randomly generated nucleotied sequence\"\"\"\n",
    "    return ''.join(random.choice('ACTG') for _ in range(length))\n",
    "\n",
    "def non_sequence(length):\n",
    "    \"\"\"Returns a randomly generated sequence with N's\"\"\"\n",
    "    return ''.join(random.choice('ACTGNNN') for _ in range(length))\n",
    "\n",
    "def constrained_sum_sample_pos(n, total):\n",
    "    \"\"\"Return a randomly chosen list of n positive integers summing to total.\n",
    "    Each such list is equally likely to occur.\"\"\"\n",
    "\n",
    "    dividers = sorted(random.sample(xrange(1, total), n - 1))\n",
    "    return [a - b for a, b in zip(dividers + [total], [0] + dividers)]\n",
    "\n",
    "def gen_qscore(length, score):\n",
    "    \"\"\"Return a fake qscore of length length\"\"\"\n",
    "    myScore = constrained_sum_sample_pos(length,score)\n",
    "    return ''.join([chr(i+33) for i in myScore])\n",
    "\n",
    "def gen_ID(alpha):\n",
    "    \"\"\"Return an ID for the fastq file five characters long using the\n",
    "    supplied alphabet\"\"\"\n",
    "    return ''.join(random.choice(alpha) for _ in range(5))\n",
    "\n",
    "def NoN(s):\n",
    "    \"\"\"Return the longest string not containing N's\"\"\"\n",
    "    s1 = \"NGNTAGAAAAAAAAAAAAAAAAAAAAAAAGTAAAAACAAAGN\"\n",
    "    s2 = \"GAGAAAAAAAAAAAAAAAAAAAAAAAGTAAAAACAAAGNNNNNNNNNNNNNNN\"\n",
    "    s3 = \"GAGAAAAAAAAAAAAAAAAAAAAAAAGTAAAAACAAAG\"\n",
    "    s4 = \"NNNNNGAGAAAAAAAAAAAAAAAAAAAAAAAGTAAAAACAAAG\"\n",
    "    s5 = \"NNNNNNNNNNNNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNN\"\n",
    "    \n",
    "    s=[s1,s2,s3,s4,s5]\n",
    "    \n",
    "    max_end = 0\n",
    "    max_start = 0;\n",
    "    current_end = 0;\n",
    "    current_start = 0;\n",
    "\n",
    "    for c in range(0, len(s)):\n",
    "        if (s[c] == 'N'):\n",
    "            if c - current_start + 1 > max_end - max_start:\n",
    "                max_end = c;\n",
    "                max_start = current_start;\n",
    "                current_start = c + 1;\n",
    "        \n",
    "    if c - current_start > max_end - max_start:\n",
    "        max_end = c + 1\n",
    "        max_start = current_start\n",
    "    return s[max_start:max_end]\n",
    "\n",
    "def return_binary(st):\n",
    "    # converts string to binary\n",
    "     return ' '.join(format(x, 'b') for x in bytearray(st))\n",
    "        \n",
    "def return_reverse_complement(seq):\n",
    "    # returns the reverse complement of a sequence\n",
    "    seq1 = 'ATCGTAGCatcgtagc'\n",
    "    seq_dict = {seq1[i]: seq1[i + 4]\n",
    "                for i in range(16) if i < 4 or 8 <= i < 12}\n",
    "    return \"\".join([seq_dict[base] for base in reversed(seq)])\n",
    "\n",
    "def gen_fake_fastq(readLength, qScore):  \n",
    "    \"\"\"Returns fake randomly generated fastq data\n",
    "    as a dictionary\"\"\"\n",
    "    alphabet = \"'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz0123456789'\"\n",
    "    store={}\n",
    "    id_Strings=[]\n",
    "    seq_R1=[]\n",
    "    seq_R2=[]\n",
    "    q_R1=[]\n",
    "    q_R2=[]\n",
    "    \n",
    "    id_Strings.append(\"@\"+gen_ID(alphabet))\n",
    "    seq_R1.append(gen_sequence(readLength))\n",
    "    seq_R2.append(gen_sequence(readLength))\n",
    "    q_R1.append(gen_qscore(readLength,qScore))\n",
    "    q_R2.append(gen_qscore(readLength,qScore))\n",
    "        \n",
    "    store['ID']=id_Strings\n",
    "    store['Seq_R1']=seq_R1\n",
    "    store['Seq_R2']=seq_R2\n",
    "    store['Q_R1']=q_R1\n",
    "    store['Q_R2']=q_R2\n",
    "    \n",
    "    return pd.DataFrame(store)\n",
    "\n",
    "def trim_adaptor(seq, adaptor, num_errors, right_side=True):\n",
    "    \"\"\"\n",
    "    The Biopython library contains a Python local alignment function suitable\n",
    "    for quick alignment of short regions. We use this to align an adaptor \n",
    "    region to a sequence and calculate the number of differences in the \n",
    "    aligned region. To handle a common case where we can find the exact\n",
    "    adaptor sequence, we first do an string match. \n",
    "    This avoids the alignment overhead in many cases:\n",
    "    \"\"\"\n",
    "\n",
    "    gap_char = '-'\n",
    "    exact_pos = str(seq).find(adaptor)\n",
    "    if exact_pos >= 0:\n",
    "        seq_region = str(seq[exact_pos:exact_pos+len(adaptor)])\n",
    "        adapt_region = adaptor\n",
    "    else:\n",
    "        seq_a, adaptor_a, score, start, end = pairwise2.align.localms(str(seq),\n",
    "                str(adaptor), 5.0, -4.0, -9.0, -0.5,\n",
    "                one_alignment_only=True, gap_char=gap_char)[0]\n",
    "        adapt_region = adaptor_a[start:end]\n",
    "        seq_region = seq_a[start:end]\n",
    "    matches = sum((1 if s == adapt_region[i] else 0) for i, s in\n",
    "            enumerate(seq_region))\n",
    "    # too many errors -- no trimming\n",
    "    if (len(adaptor) - matches) > num_errors:\n",
    "        return seq\n",
    "    # remove the adaptor sequence and return the result\n",
    "    else:\n",
    "        return _remove_adaptor(seq, seq_region.replace(gap_char, \"\"),\n",
    "                right_side)\n",
    "\n",
    "def _remove_adaptor(seq, region, right_side=True):\n",
    "    \"\"\"\n",
    "    Remove an adaptor region and all sequence to the right or left.\n",
    "    If the alignment contains fewer than the specified number of\n",
    "    errors we’ve found an adaptor and remove it, returning the trimmed\n",
    "    sequence. The attribute right_side allows us to specify whether\n",
    "    the trimming should be done on the right (3′ end) or the left (5′ end):\n",
    "    \"\"\"\n",
    "    if right_side:\n",
    "        try:\n",
    "            pos = seq.find(region)\n",
    "        # handle Biopython SeqRecords\n",
    "        except AttributeError:\n",
    "            pos = seq.seq.find(region)\n",
    "        return seq[:pos]\n",
    "    else:\n",
    "        try:\n",
    "            pos = seq.rfind(region)\n",
    "        # handle Biopython SeqRecords\n",
    "        except AttributeError:\n",
    "            pos = seq.seq.rfind(region)\n",
    "        return seq[pos+len(region):]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Here is an example script using this function along with Biopython\n",
    "to trim reads from a FASTQ format file. Each read is analyzed to\n",
    "determine if it contains the adaptor, with up to 2 errors.\n",
    "If the adaptor is identified and trimmed, the final useful\n",
    "read is written to a FASTA format output file.\n",
    "\"\"\"\n",
    "\n",
    "from __future__ import with_statement\n",
    "import sys\n",
    "import os\n",
    " \n",
    "from Bio import SeqIO\n",
    "# The in_file var will need to be changed\n",
    "in_file = \"/Users/ChrisM/Documents/workspace/iBest/fastaFiles/rawData/NTC_R1.fastq\"\n",
    "adaptor_seq = \"ACACTCTTTCCCTACACGACGCTCTTCCGATCT\"\n",
    "reverse_complement = return_reverse_complement(adaptor_seq)\n",
    "num_errors = 2\n",
    "base, ext = os.path.splitext(in_file)\n",
    "out_file = \"%s-trimmed.fa\" % (base)\n",
    "\n",
    "with open(in_file) as in_handle:\n",
    "    with open(out_file, \"w\") as out_handle:\n",
    "        for rec in SeqIO.parse(in_handle, \"fastq\"):\n",
    "            trim = trim_adaptor(rec.seq, adaptor_seq, num_errors)\n",
    "            if len(trim) > 0 and len(trim) < len(rec):\n",
    "                rec.letter_annotations = {}\n",
    "                rec.seq = trim\n",
    "                SeqIO.write([rec], out_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
